sys/net/application_layer/gcoap/gcoap_helper.c

```diff
@@ -0,0 +1,218 @@
+ * @ingroup     net_gcoap_helper
+ * @{
+ *
+ * @file
+ * @brief       Helper functions for GCoAP
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include "net/gcoap_helper.h"
+#include "net/utils.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#define COAP_REQ_CONFIRMABLE    (1 << 0)
+
+struct coap_request_ctx {
+    mutex_t lock;
+    const void *payload;
+    const size_t payload_len;
+    void *dst;
+    size_t dst_len;
+    const size_t dst_len_max;
+    const char *path;
+    int code;
+    uint8_t method;
+    uint8_t flags;
+};
+
+/*
+ * Response callback.
+ */
+static void _resp_handler(const gcoap_request_memo_t *memo, coap_pkt_t* pdu,
+                          const sock_udp_ep_t *remote)
+{
+    (void)remote;       /* not interested in the source currently */
+
+    struct coap_request_ctx *ctx = memo->context;
+
+    if (memo->state == GCOAP_MEMO_TIMEOUT) {
+        DEBUG("gcoap: timeout for msg ID %02u\n", coap_get_id(pdu));
+        ctx->code = -ETIME;
+        goto out;
+    }
+    else if (memo->state == GCOAP_MEMO_ERR) {
+        DEBUG("gcoap: error in response\n");
+        ctx->code = -EBADMSG;
+        goto out;
+    }
+
+    coap_block1_t block;
+    char *class_str = (coap_get_code_class(pdu) == COAP_CLASS_SUCCESS)
+                            ? "Success" : "Error";
+    DEBUG("gcoap: response %s, code %1u.%02u\n", class_str,
+                                                 coap_get_code_class(pdu),
+                                                 coap_get_code_detail(pdu));
+    ctx->code = coap_get_code(pdu);
+
+    if (pdu->payload_len) {
+        if (ctx->dst_len + pdu->payload_len > ctx->dst_len_max) {
+            DEBUG("gcoap: not enough memory in destination buffer\n");
+            ctx->code = -ENOMEM;
+            goto out;
+        }
+        memcpy((uint8_t *)ctx->dst + ctx->dst_len, pdu->payload, pdu->payload_len);
+        ctx->dst_len += pdu->payload_len;
+    }
+
+    /* ask for next block if present */
+    if (coap_get_block2(pdu, &block)) {
+
+        if (block.blknum == 0) {
+            DEBUG("--- blockwise start ---\n");
+        }
+
+        if (block.more) {
+            unsigned msg_type = coap_get_type(pdu);
+
+            gcoap_req_init(pdu, (uint8_t *)pdu->hdr, CONFIG_GCOAP_PDU_BUF_SIZE,
+                           ctx->method, ctx->path);
+
+            if (msg_type == COAP_TYPE_ACK) {
+                coap_hdr_set_type(pdu->hdr, COAP_TYPE_CON);
+            }
+            block.blknum++;
+            coap_opt_add_block2_control(pdu, &block);
+
+            int len = coap_opt_finish(pdu, COAP_OPT_FINISH_NONE);
+            gcoap_req_send((uint8_t *)pdu->hdr, len, remote,
+                           _resp_handler, memo->context);
+        }
+        else {
+            DEBUG("--- blockwise complete ---\n");
+        }
+    }
+
+out:
+    mutex_unlock(&ctx->lock);
+}
+
+static void coap_request(const sock_udp_ep_t *remote, struct coap_request_ctx *ctx)
```

Comment:
Can you also add support for setting the block size?  The default 16 bytes is for some use cases a little bit too small.

You could use `coap_block_object_init()`.
http://doc.riot-os.org/group__net__nanocoap.html#gadabe7197cbe80bf17fea0ecf5ee9bf56

E.g.:
```c
coap_block1_t block;
coap_block_object_init(&ctx->block, 0, ctx->blksz, 0);
coap_opt_add_block2_control(&pdu, &block);
int len = coap_opt_finish(&pdu, COAP_OPT_FINISH_XXX);
```